src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4355,47 +4364,54 @@ private static TypeWithState GetNullCoalescingResultType(TypeWithState rightResu
                 resultType = BestTypeInferrer.InferBestTypeForConditionalOperator(consequencePlaceholder, alternativePlaceholder, _conversions, out _, ref discardedUseSiteInfo);
             }
 
-            NullableFlowState resultState;
-            if (resultType is null)
-            {
-                resultType = node.Type?.SetUnknownNullabilityForReferenceTypes();
-                resultState = consequenceRValue.State.Join(alternativeRValue.State);
 
-                var resultTypeWithState = TypeWithState.Create(resultType, resultState);
+            resultType ??= node.Type?.SetUnknownNullabilityForReferenceTypes();
 
-                if (consequence != originalConsequence)
+            NullableFlowState resultState;
+            if (!wasTargetTyped)
+            {
+                if (resultType is null)
                 {
-                    TrackAnalyzedNullabilityThroughConversionGroup(resultTypeWithState, (BoundConversion)originalConsequence, consequence);
+                    // This can happen when we're inferring the return type of a lambda. For this case, we don't need to do any work,
+                    // the unconverted conditional operator can't contribute info. The conversion that should be on top of this
+                    // can add or remove nullability, and nested nodes aren't being publicly exposed by the semantic model.
+                    Debug.Assert(node is BoundUnconvertedConditionalOperator);
+                    Debug.Assert(_returnTypesOpt is not null);
+                    resultState = default;
                 }
-
-                if (alternative != originalAlternative)
+                else
                 {
-                    TrackAnalyzedNullabilityThroughConversionGroup(resultTypeWithState, (BoundConversion)originalAlternative, alternative);
+                    var resultTypeWithAnnotations = TypeWithAnnotations.Create(resultType);
+
+                    TypeWithState convertedConsequenceResult = ConvertConditionalOperandOrSwitchExpressionArmResult(
+                        originalConsequence,
+                        consequence,
+                        consequenceConversion,
+                        resultTypeWithAnnotations,
+                        consequenceRValue,
+                        consequenceState,
+                        consequenceEndReachable);
+
+                    TypeWithState convertedAlternativeResult = ConvertConditionalOperandOrSwitchExpressionArmResult(
+                        originalAlternative,
+                        alternative,
+                        alternativeConversion,
+                        resultTypeWithAnnotations,
+                        alternativeRValue,
+                        alternativeState,
+                        alternativeEndReachable);
+
+                    resultState = convertedConsequenceResult.State.Join(convertedAlternativeResult.State);
                 }
             }
             else
             {
```

Comment:
If I understood correctly, the strategy is to postpone/delay analysis of the branch when visiting the target-typed conditional/switch expression, but come back to it when we visit the conversion that provides a target-type.
We do this by storing information from `VisitConditionalOperator` and passing it up to `VisitConversion`, so that we can restore the proper state before each branch.

nit: if that's correct, it may be good to add a comment.

Would it be possible for `VisitConversion` to pass information (namely the target type) down to `VisitConditionalOperator` instead? If possible, that would seem simpler and less information to store.

Reply:
> Would it be possible for VisitConversion to pass information (namely the target type) down to VisitConditionalOperator instead? If possible, that would seem simpler and less information to store, and maybe we can factor with code calling ConvertConditionalOperandOrSwitchExpressionArmResult above.

No, or more precisely yes, but _everywhere_ that one of these things could be would need to have a ton of special code added. Everywhere in the walker that we call `RemoveConversion` would need to be updated to specially detect if the operand was target-typed, and if so do something special. That's a lot of peanut butter spread across the walker and I'm not certain that we'd catch everything. This has the advantage of working with our existing pattern.

> the strategy is to postpone/delay analysis of the branch when visiting the target-typed conditional/switch expression

Not quite. We delay analysis of the _conversion_ on the branch. Not on the branch itself.

---
In reply to: [595629506](https://github.com/dotnet/roslyn/pull/51791#discussion_r595629506) [](ancestors = 595629506)

Reply:
I'll add some more comments to try and convey this.

---
In reply to: [597050103](https://github.com/dotnet/roslyn/pull/51791#discussion_r597050103) [](ancestors = 597050103,595629506)

Reply:
Thanks. Additional comments will be useful

---
In reply to: [597050322](https://github.com/dotnet/roslyn/pull/51791#discussion_r597050322) [](ancestors = 597050322,597050103,595629506)